### ISSUE TITLE
Allow assignment of hot corner to specific corner and monitor(s)

### ISSUE THREAD

**Issue by amxmln** [home-sweet-gnome/dash-to-panel#216]
Since the last update, the hot corner for the activities overview of my main screen is no longer working, it only works on my secondary screen which is positioned to the left of my main screen. Disabling the extension fixes the issue, but it appears again as soon as the extension is re-enabled.

**Comment by jderose9** [comment]
This was intentional as it seemed undesirable to have a floating hot corner that trapped the mouse in between the two monitors when moving between the displays.

Do you mind to try the custom corner extension and see if you are able to add it back that way?
https://extensions.gnome.org/extension/1037/customcorner/

Perhaps it would be best to add a config setting for this..

**Comment by weltenweber** [comment]
Actually it is precisely that "trapping" of the mouse that is missing from the extension you suggested, which basically makes it unusable since it becomes rather hit and miss whether the hot corner fires or not.

I think an option for it in the config would be for the best—it seems counterproductive to have to add an extension to get default functionality back even if it were working as intended.

**Comment by jderose9** [comment]
The hot corner should never fire if there is a monitor to the left. Are you saying that it sometimes does?

**Comment by weltenweber** [comment]
No, not with only Dash To Panel activated, sorry for the confusion. I meant only when I turn on the Custom Corner extension.

**Comment by jderose9** [comment]
Oh, I see.

I don't really agree with the idea that default functionality was removed. The default hot corner was in the corner of the panel. After moving the panel to the bottom, the hot corner was essentially "floating" between monitors. Does moving the hot corner to the bottom along with the panel make more sense?

**Comment by weltenweber** [comment]
I never saw the hot corner as an attachment to the panel, I always have the top left corner set up to show an overview no matter if I’m using GNOME or Pantheon, but that is my personal preference.

I think keeping the hot corner in the corner of the panel (i.e. moving it to the bottom) would definitely be preferable to having to go all the way to the far left of my secondary screen, however, my cursor (and most people’s in nations where text is read from top to bottom and left to right) naturally is mostly within the top left quadrant of the screen.

Maybe this is all just me and my habits though, so I’d still think having it as an option in the config would be the best solution.

**Comment by jderose9** [comment]
Ok, sounds like we need to have a couple of config settings. Something along the lines of choosing which corner to assign and whether it fires on all monitors, primary monitor only, or furthest monitor only.

**Comment by bentrop** [comment]
> This was intentional as it seemed undesirable to have a floating hot corner that trapped the mouse in between the two monitors when moving between the displays.

I was just going to file a new bug report, but it looks like this "intentional change" in 2017 may be the cause of the issue I'm now facing in 2020.

I'm a first time Dash to Panel user and on Manjora with a dual monitor set-up with the **primary** monitor on the **right** and the secondary screen extending it to the left. 

On a vanilla Gnome install without extensions that set up is no problem. The "activities" hot corner works on **both** screens (top left) as intended.

Enabling Dash to Panel breaks this behavior and the hot corner now only works on the leftmost screen, **not** the primary screen.

The suggested "Custom Corner" extension does not actually allow me to fix this issue.

I'm still somewhat torn if this should be a bug report since this literally breaks Gnome default behavior and Gnome's intended workflow. Not being able to have the activities hot corner on my **primary** screen because of the dash to panel extension doesn't make much sense to me and three years later there still isn't a setting to revert Gnome back to its - correct in my eyes and for my set-up - default behavior. At least not a setting I could find.

I don't require much customization but could we maybe just have a "don't screw up Gnome's default hot corners" setting somewhere?

**Comment by charlesg99** [comment]
> this literally breaks Gnome default behavior and Gnome's intended workflow

Hmm, but not setting it's panel to the bottom? :) If you want the closest to gnome-shell's intended workflow, the solution is simple, set dash-to-panel to the Top position, then you'll get the hot corner on your right monitor. Gnome-shell's behavior is to set a top left hot corner if this is the primary monitor (where there always is a top panel), or if there is no monitor directly above or to the left. We are currently doing exactly that.

If you want to keep the panel at the bottom, but need top left hot corners on every monitors, you could run this command, then restart gnome-shell:
```
sed -i "s/|| panelPosition == St.Side.BOTTOM || panelPosition == St.Side.RIGHT//" ~/.local/share/gnome-shell/extensions/dash-to-panel@jderose9.github.com/panelManager.js
```
Thanks!

**Comment by bentrop** [comment]
> Gnome-shell's behavior is to set a top left hot corner if this is the primary monitor (where there always is a top panel), or if there is no monitor directly above or to the left. We are currently doing exactly that.

Mhhh, maybe this is a bug then, because that is exactly my problem: you are **not** currently doing exactly that.

My **right** monitor is my **primary** monitor. As soon as I enable Dash to Panel, there is no more hot corner _anywhere_ on my **primary** monitor. It's now only functional on my secondary screen, which happens to be to the left of the primary screen.

The "Main Panel" of Dash to Dock (set to be on the primary monitor, default setting) and the activities hot corner are now on two entirely separate screens and I found no setting to fix that.

Is this indeed intended behavior, which I would find rather odd to say the least, or a bug then?

**Comment by charlesg99** [comment]
tl;dr As I said, just set dash to panel to the Top position or run the command I provided to have it the way you want. 

I understood your setup and problem the first time. What I'm saying is that gnome-shell doesn't add a hot corner on a monitor that has no top panel and has a monitor directly above or to the left. You can easily verify that on a gnome-shell session by setting your primary monitor to the left, you won't have a hot corner on your right monitor. 

Now, we can speculate all day about what would happen on your multi-monitor gnome-shell session if their panel was at the bottom, but the only way to get a clear answer would be to ask them. I'm sure they never thought of it since there always is a top panel on the primary monitor.

Thanks!

**Comment by bentrop** [comment]
Thank you so much for taking the time to explain this @charlesg99 
I believe I finally understand the cause of the issue and that it's Gnome Shell's doing, not intentional behavior of Dash to Panel.

I still feel like this probably should be addressed within Dash to Panel regardless, since the removal of all hot corners from the primary screen can't possibly be desirable behavior - and that's exactly what's happening every time when enabling Dash to Panel, with its default settings, on a multi-monitor setup where the primary screen isn't all the way to the left.

**Comment by charlesg99** [comment]
Did you try running the command I provided, then restart gnome-shell? Should do the trick to keep the hot corner on the primary monitor, no matter the panel position.

If you can think of a good setting label/name and a sensible location for it (within our cramped settings dialog), I'll be happy to put it in to close this issue. Thanks!

**Comment by bentrop** [comment]
Since Dash to Panel comes preinstalled on Manjaro Gnome, I had to change it to /usr/share/ but, yes, it works very well and returns the hot corners to the behavior before enabling Dash to Panel - at least on my set-up.

I feel like it would likely fit in best under "Fine-Tune" in the box with "Keep original gnome-shell dash" since that box is all about preserving the status quo.
Or in "Behavior" underneath "Show Activities Button" because it's about the "Activities" hot corner.

I'm not sure if I understand what precisely the command does. Does it enable the hot corner on _every_ single screen, even on multi-monitor set-up. Or does it simply restore the hot corner on the primary screen, even if there is no top-panel?

If it's the latter, it should be titled something like "(Force) enable [Activities] hot corner on primary display" and it likely should be enabled by default, since that mimics Gnome shell's default behavior and is quite obviously the more desirable state for just about any user that would be affected by this setting. The "(force)" part might be necessary since disabling the setting wouldn't actually disable the hot corner on most set-ups. The setting would be 100% useless on single-screen setups or set-ups where the primary display is also the left-most display.

That's why, frankly, I'm not even sure if it needs to be a _setting_ at all. It's hard to think of a user in my position who would actually want the current behavior. It's already possibly to disable the hot corner entirely in various ways for anyone who doesn't like it, just moving the hot corner away to a secondary screen sounds ... odd and not like it should be default or desired behavior. Maybe it's a prime candidate for your first "Advanced Option", but again: since it shouldn't effect that majority of user at all and would only affect a small percentage in a positive way, I'm not sure if needs to be a toggle at all. (Maybe it does, though. Relevant xkcd: https://xkcd.com/1172/ )

If it enables the hot corner on **all** screens, not just the leftmost and the primary display, just call it:
"Enable [Activities] hot corner on all displays". Sounds like a neat feature to have. The possible positions would be the same.

**Comment by jderose9** [comment]
@bentrop 
In contrast, I can't imagine why anyone would want to do what you are describing instead of the current behavior - why would I want some annoying hot CORNER popping up when all I am trying to do is move my mouse seamlessly across my desktop? Wouldn't it make more sense to go into the actual physical corner of the displays that make up my desktop area? 

So, I guess we will just have to settle on people having different preferences? :)

That's probably the crux of it, though - people who think of their workflow revolving around the primary monitor and the secondary monitors are ancillary, and those who think of their monitors as a single seamless desktop. So in that vein, maybe "Move hot corner to physical corner" or something along those lines make sense for labelling?

Just to be clear - nobody is proposing we should have a hot corner on every display, correct?

**Comment by bentrop** [comment]
> why would I want some annoying hot CORNER popping up when all I am trying to do is move my mouse seamlessly across my desktop?

Why are you moving your mouse across two displays within 2 pixels of the top of your screen?

> Wouldn't it make more sense to go into the actual physical corner of the displays that make up my desktop area?

Sure. But if that's what makes sense to you, then why would you not set _that_ screen as your primary screen?

> That's probably the crux of it, though - people who think of their workflow revolving around the primary monitor and the secondary monitors are ancillary, and those who think of their monitors as a single seamless desktop.

Again, if you see it as a single desktop, why wouldn't your primary screen _not_ be the one where you would also want to interact with Gnome's "activities"?
Also: Seeing it as a single desktop may make sense with identically sized, slim bezel screens.
If you have different sized screens with different functions, suddenly not having easy access to Gnome's "activities" on your _primary_ screen just because you enabled an otherwise very nice and functional extension is kind of a bummer.
Also: Dash to Panel _itself_ follows the "they are individual screens not a single desktop" school of thought by providing a primary panel and less functional secondary clones. Why would you only want access to the activities corner on a screen with such a clone?

> So, I guess we will just have to settle on people having different preferences? :)

Sure, but those with single monitors would be unaffected anyways. Those with their primary screen all the way to the left would also be fully unaffected. It's only freaks like me why extend their screens to the left who would even notice the change.
I fail to see why those with multi monitors who see them as one big, fully homogeneous desktop wouldn't pretty much automatically fall into the second category. Heck, your primary "Dash to Panel" doesn't even have to be on your primary screen. So just for the freak occurrence that someone out of this group would be affected by this change and would now be annoyed that the hot corner returns back to where Gnome would have put it without Dash to Panel anyways - the solution would be incredibly simple: set the leftmost screen as your primary and put the primary "Dash to Panel" panel wherever you want it.

> Just to be clear - nobody is proposing we should have a hot corner on every display, correct?
All I want is Gnome's default behavior even when using Dash to Panel: have an activities hot corner on my primary _and_ leftmost display. That's currently all displays for me anyways. 99.9% of all interaction with the activities screen would happen on my primary screen, so I donÄt really care about the other screen's hot corners.

**Comment by jderose9** [comment]
@bentrop *I* am affected by the change you are proposing. So there is at least one person out there. :) 

There are not only users with one or two displays in a horizontal configuration. There are also users with a top and bottom display, or three displays with the primary taskbar in the middle.. If I so choose to have a 9x9 grid of display, should I be forced to have my taskbar in the top left display?

**Comment by charlesg99** [comment]
Hey guys! I just added the option to the master branch, hopefully that should settle the issue. It can be found under the "Fine-Tune" tab, as it is a way to duplicate the default gnome-shell's behavior of having a hot corner on the primary monitor, no matter where the panel is (the other monitors are unaffected). It is disabled by default, so this shouldn't affect existing users. 

Thanks!

**Comment by charlesg99** [comment]
> Since Dash to Panel comes preinstalled on Manjaro Gnome

Hmm that's new? I'll have to check it out :) @jderose9 did you know about this?

**Comment by jderose9** [comment]
@charlesg99 Nope, news to me. Go Manjaro :)

**Comment by bentrop** [comment]
> If I so choose to have a 9x9 grid of displays, should I be forced to have my taskbar in the top left display?

My guess is, that a single hot corner would likely be the very least of your worries in such a scenario.
But just to play your game: which one would be your primary screen in this theoretical scenario?
To be frank, I still wouldn't know how to answer it, what does Gnome shell do in that case? Just make sure that's what happens with Dash to Panel active as well ...

My point is that the average user would expect hot corner behavior to remain unchanged when enabling a seemingly unrelated extension like Dash to Panel - and that's currently not what happens.

> Hey guys! I just added the option to the master branch, hopefully that should settle the issue.

Woohoo! Thanks! (I still think it should be enabled by default, but ultimately I'm happy either way).

> Hmm that's new? I'll have to check it out :)

It's part of the manjaro-gnome-assets package. Installed and even enabled by default, in combination with "Arc Menu". It also installs Dash-to-Dock, but only enables Dash-to-Panel. I think this has been around for a while. Manjaro offers an option to revert to vanilla Gnome upon install, though.

**Comment by jderose9** [comment]
@charlesg99 Are we good with this implementation? Or is there reason to allow more specific assignment of the hot corner location (bottom-left, top-right, etc)?
